### PR TITLE
Compatibility with molinillo 0.6

### PR DESCRIPTION
**Pull request by thommay**
Fixes the bug where we'd arbitrarily ignore dependencies

### CODE REVIEW

**Comment by thommay** [comment]
Fixes: berkshelf/berkshelf#1723 and #67

File: lib/solve/dependency.rb [hunk 1]
```diff
@@ -27,12 +27,14 @@ def initialize(artifact, name, constraint = Semverse::DEFAULT_CONSTRAINT)
...
       "#{name} (#{constraint})"
     end
+    alias :inspect :to_s
 
     # @param [Object] other
     #
     # @return [Boolean]
     def ==(other)
       other.is_a?(self.class) &&
+        name == other.name &&
```
Comment (thommay) [inline lib/solve/dependency.rb]: Previously, the solver was checking to ensure that we'd not seen the requirement to use a dependency before. ( https://github.com/CocoaPods/Molinillo/blob/master/lib/molinillo/resolution.rb#L754 ) . We'd compare two dependencies, and if they both had the same *parent* - the artifact - and the same constraint, they'd pass, regardless of whether the *name* of the dependency was the same.

File: lib/solve/ruby_solver.rb [hunk 2]
```diff
@@ -159,9 +162,29 @@ def search_for(dependency)
     # Callback required by Molinillo
     # @return [Boolean]
     def requirement_satisfied_by?(requirement, activated, spec)
-      requirement.constraint.satisfies?(spec.version)
+      version = spec.version
+      return false unless requirement.constraint.satisfies?(version)
+      shared_possibility_versions = possibility_versions(requirement, activated)
```
Comment (thommay) [inline lib/solve/ruby_solver.rb]: molinillo 0.6 allows us to check all the vertices of the dependency graph at once, letting us resolve more quickly.

**Comment by thommay** [comment]
I can't reproduce that chefstyle failure.

File: lib/solve/ruby_solver.rb [hunk 3]
```diff
@@ -159,9 +162,29 @@ def search_for(dependency)
...
+      return false if !shared_possibility_versions.empty? && !shared_possibility_versions.include?(version)
+      true
     end
 
+    def possibility_versions(requirement, activated)
```
Comment (coderanger) [inline lib/solve/ruby_solver.rb]: YARD doc comment for the future?

**Pull request merged**